File: src/main/adaptive-rate-limiter.ts
const abortableDelay = (ms: number, signal: AbortSignal): Promise<void> => new Promise((resolve, reject) => {
  const onAbort = () => {
    clearTimeout(timer);
    reject(new DOMException("查询已取消", "AbortError"));
  };
  const timer = setTimeout(() => {
    signal.removeEventListener("abort", onAbort);
    resolve();
  }, ms);
  if (signal.aborted) onAbort();
  else signal.addEventListener("abort", onAbort, { once: true });
});

export class AdaptiveRateLimiter {
  private readonly baseDelayMs: number;
  private readonly maxDelayMs: number;
  private readonly min429DelayMs: number;
  private currentDelayMs: number;
  private nextRequestAt = 0;

  constructor(baseDelayMs = 350, maxDelayMs = 3000, initialDelayMs = baseDelayMs, min429DelayMs = 1500, delayFirstRequest = false) {
    this.baseDelayMs = baseDelayMs;
    this.maxDelayMs = maxDelayMs;
    this.min429DelayMs = min429DelayMs;
    this.currentDelayMs = Math.min(maxDelayMs, Math.max(baseDelayMs, initialDelayMs));
    if (delayFirstRequest) this.nextRequestAt = Date.now() + this.currentDelayMs;
  }

  get delayMs(): number {
    return Math.round(this.currentDelayMs);
  }

  async wait(signal: AbortSignal): Promise<void> {
    const now = Date.now();
    const scheduledAt = Math.max(now, this.nextRequestAt);
    this.nextRequestAt = scheduledAt + this.currentDelayMs;
    const remaining = scheduledAt - now;
    if (remaining > 0) await abortableDelay(remaining, signal);
  }

  onSuccess(): void {
    if (this.currentDelayMs > this.baseDelayMs) {
      this.currentDelayMs = Math.max(this.baseDelayMs, Math.ceil(this.currentDelayMs * 0.8));
    }
  }

  onBackpressure(status: number): void {
    const minimum = status === 429 ? this.min429DelayMs : this.baseDelayMs;
    this.currentDelayMs = Math.min(this.maxDelayMs, Math.max(minimum, Math.ceil(this.currentDelayMs * 1.7)));
  }

  onNetworkError(): void {
    this.currentDelayMs = Math.min(this.maxDelayMs, Math.ceil(this.currentDelayMs * 1.35));
  }
}

File: android/app/src/main/java/cn/ldxp/sourcebrowser/android/network/AdaptiveRateLimiter.java
package cn.ldxp.sourcebrowser.android.network;

import java.util.concurrent.atomic.AtomicBoolean;

public final class AdaptiveRateLimiter {
    private final int baseDelayMs;
    private final int maxDelayMs;
    private final int minimum429Ms;
    private int delayMs;
    private long nextRequestAt;

    public AdaptiveRateLimiter(int baseDelayMs, int maxDelayMs, int initialDelayMs, int minimum429Ms, boolean delayFirst) {
        this.baseDelayMs = baseDelayMs;
        this.maxDelayMs = maxDelayMs;
        this.minimum429Ms = minimum429Ms;
        delayMs = Math.max(baseDelayMs, Math.min(maxDelayMs, initialDelayMs));
        if (delayFirst) nextRequestAt = System.currentTimeMillis() + delayMs;
    }

    public int delayMs() {
        synchronized (this) { return delayMs; }
    }

    public void waitTurn(AtomicBoolean cancelled) throws ApiException {
        long wait;
        synchronized (this) {
            long now = System.currentTimeMillis();
            long scheduled = Math.max(now, nextRequestAt);
            nextRequestAt = scheduled + delayMs;
            wait = scheduled - now;
        }
        while (wait > 0) {
            if (cancelled.get()) throw new ApiException("查询已取消", 0, 0, false);
            long slice = Math.min(100, wait);
            try { Thread.sleep(slice); } catch (InterruptedException error) {
                Thread.currentThread().interrupt();
                throw new ApiException("查询已取消", 0, 0, false);
            }
            wait -= slice;
        }
    }

    public synchronized void onSuccess() {
        if (delayMs > baseDelayMs) delayMs = Math.max(baseDelayMs, (int) Math.ceil(delayMs * 0.8));
    }

    public synchronized void onBackpressure(int status) {
        int minimum = status == 429 ? minimum429Ms : baseDelayMs;
        delayMs = Math.min(maxDelayMs, Math.max(minimum, (int) Math.ceil(delayMs * 1.7)));
    }

    public synchronized void onNetworkError() {
        delayMs = Math.min(maxDelayMs, (int) Math.ceil(delayMs * 1.35));
    }
}
